### ISSUE TITLE
Normalize and link affiliations

### ISSUE THREAD

**Issue by michamos** [inspirehep/inspirehep#1530]
As explained in #1260, we often receive unnormalized/unlinked affiliations on Literature records.

We need to implement a way to normalize raw affiliations, and systematically link all unlinked affiliations (coming from authors.xml or from this raw affiliation normalization).

**Comment by StellaCh** [comment]
Test ES with Micha's dataset

**Comment by michamos** [comment]
Dataset: see https://github.com/inspirehep/inspire-next/issues/1875#issuecomment-273832266

**Comment by MJedr** [comment]
The accuracy for the solution proposed by Paweł in #1260 on Micha's dataset (with the new mapping and splitting names) was 54% (measuring only the best matches).

I agree that splitting the input to the name part and address part is crucial to get the better accuracy, but it is necessary to find a good set of rules for that. For that reason, I tried to create a better way of splitting than in Paweł's solution based on the length of the split affiliation and the best approach for my tests was to check if there's a keyword suggesting the name. I tried also the recognition of the address part with named entity recognition from `nltk` library but it wasn't great (for example it was considering `University` as an address).

Also, for Micha dataset the `_ins_fingerprint` from Paweł's mapping wasn't the best solution (there was too many information possible to match). Instead, what worked better for me wa to do the multimatch query with increased weight for the legacy_ICN field (it was especially helpful for the names such as `U. of Yale`).

The best score that I got from my tests was 62,7% accuracy using the script below:
```
def benchmark(data):
    correct = 0
    name_regex = re.compile(
        "Academy|Acade.*|U\.|Laboratory|Labo.*|Facul.*|Departament|Center|Physics|Technology|Research|Observat*|Insti*|^\s*[A-Z]*\s*$|Univer|Uni\..*")
    for sample in data:
        might_be_country = ''
        name_with_uni_synonyms_replaced = re.sub("(University|Uni\.|U(?=<\s))", "U.\s",  sample[1])
        name_splitted = re.split('\s-\s|,', name_with_uni_synonyms_replaced)
        address_part = []
        name_part = []
        if len(name_splitted) == 1:
            might_be_country_found = re.findall("\([A-Z].*\)", name_splitted[0])
            if might_be_country_found:
                might_be_country = might_be_country_found[0].strip('\(\)')
            name_part = name_splitted
        if len(name_splitted) >= 2:
            name_part = [name_splitted[0]]
            for token in name_splitted[1:]:
                if not bool(re.findall(name_regex, token)):
                    address_part.append(token)
                else:
                    name_part.append(token)
        name_part = " ".join(name_part)
        address_part = " ".join(address_part)
        query = Q("multi_match", query=r"{}".format(name_part), fields=["legacy_ICN^3", "ICN", "extra_words", "institution_hierarchy.acronym"],
                  type="most_fields")
        if name_part and address_part:
            query = query & Q("match", _address_fingerprint=address_part)
        elif might_be_country:
            query = query & Q("match", addresses__country=might_be_country)
        result = InstitutionsSearch().query(query).execute().hits
        if result:
            if result[0].legacy_ICN == sample[0]:
                print("Hit!", result[0].legacy_ICN)
                correct += 1
            else:
                print(f"Missed, expected {sample[0]} got {result[0].legacy_ICN}")

    return correct
```

The mapping:

```
{
  "mappings": {
    "date_detection": false,
    "numeric_detection": false,
    "properties": {
      "$schema": {
        "type": "keyword"
      },
      "_address_fingerprint": {
        "type": "text",
        "analyzer": "institutions_analyzer"
      },
      "ICN": {
        "copy_to":["affautocomplete"],
        "type": "text"
      },
      "_created": {
        "type": "date"
      },
      "_updated": {
        "type": "date"
      },
      "_collections": {
        "normalizer": "lowercase_normalizer",
        "type": "keyword"
      },
      "_private_notes": {
        "properties": {
          "source": {
            "type": "keyword"
          },
          "value": {
            "type": "text"
          }
        },
        "type": "object"
      },
      "addresses": {
        "properties": {
          "cities": {
            "type": "keyword",
            "copy_to": ["_address_fingerprint"]
          },
          "country_code": {
            "type": "keyword",
            "copy_to": ["_address_fingerprint"]
          },
          "country": {
            "type": "keyword",
            "copy_to": ["_address_fingerprint"]
          },
          "latitude": {
            "type": "double"
          },
          "longitude": {
            "type": "double"
          },
          "place_name": {
            "type": "text",
            "copy_to": ["_address_fingerprint"]
          },
          "postal_address": {
            "type": "text",
            "copy_to": ["_address_fingerprint"]
          },
          "postal_code": {
            "type": "text",
            "copy_to": ["_address_fingerprint"]
          },
          "state": {
            "type": "keyword",
            "copy_to": ["_address_fingerprint"]
          }
        },
        "type": "object"
      },
      "affautocomplete": {
        "type": "text"
      },
      "affiliation_suggest": {
        "type": "completion"
      },
      "control_number": {
        "type": "integer"
      },
      "core": {
        "type": "boolean"
      },
      "deleted": {
        "type": "boolean"
      },
      "deleted_recids": {
        "type": "integer"
      },
      "deleted_records": {
        "properties": {
          "$ref": {
            "type": "text",
            "analyzer": "recid_analyzer"
          }
        },
        "type": "object"
      },
      "external_system_identifiers": {
        "properties": {
          "schema": {
            "type": "keyword"
          },
          "value": {
            "type": "keyword"
          }
        },
        "type": "object"
      },
      "extra_words": {
        "type": "text"
      },
      "historical_data": {
        "type": "text"
      },
      "inactive": {
        "type": "boolean"
      },
      "inspire_categories": {
        "properties": {
          "source": {
            "type": "keyword"
          },
          "term": {
            "type": "keyword"
          }
        },
        "type": "object"
      },
      "institution_hierarchy": {
        "properties": {
          "acronym": {
            "copy_to": ["affautocomplete"],
            "type": "text"
          },
          "name": {
            "type": "text"
          }
        },
        "type": "object"
      },
      "institution_type": {
        "type": "keyword"
      },
      "legacy_ICN": {
        "copy_to": "affautocomplete",
        "type": "text",
        "analyzer": "institutions_analyzer"
      },
      "legacy_creation_date": {
        "format": "yyyy-MM-dd||yyyy-MM||yyyy",
        "type": "date"
      },
      "name_variants": {
        "properties": {
          "source": {
            "type": "keyword"
          },
          "value": {
            "copy_to": ["affautocomplete"],
            "type": "text"
          }
        },
        "type": "object"
      },
      "new_recid": {
        "type": "integer"
      },
      "new_record": {
        "properties": {
          "$ref": {
            "type": "text",
            "analyzer": "recid_analyzer"
          }
        },
        "type": "object"
      },
      "public_notes": {
        "properties": {
          "source": {
            "type": "keyword"
          },
          "value": {
            "type": "text"
          }
        },
        "type": "object"
      },
      "related_records": {
        "properties": {
          "curated_relation": {
            "type": "boolean"
          },
          "recid": {
            "type": "integer"
          },
          "record": {
            "properties": {
              "$ref": {
                "type": "text",
                "analyzer": "recid_analyzer"
              }
            },
            "type": "object"
          },
          "relation": {
            "type": "keyword"
          },
          "relation_freetext": {
            "type": "text"
          }
        },
        "type": "nested"
      },
      "self": {
        "properties": {
          "$ref": {
            "type": "text",
            "analyzer": "recid_analyzer"
          }
        },
        "type": "object"
      },
      "self_recid": {
        "type": "integer"
      },
      "urls": {
        "properties": {
          "description": {
            "type": "text"
          },
          "value": {
            "type": "text"
          }
        },
        "type": "object"
      },
      "number_of_papers": {
        "type": "integer"
      }
    }
  },
  "settings": {
    "analysis": {
      "filter": {
          "synonym": {
            "type": "synonym",
            "synonyms": [
              "University, Uni, U => U."
            ]
          }
        },
      "analyzer": {
        "recid_analyzer": {
          "tokenizer": "recid_tokenizer"
        },
        "institutions_analyzer": {
          "tokenizer": "standard",
          "filter": [
            "asciifolding"
          ]
        }
      },
      "normalizer": {
        "lowercase_normalizer": {
          "type": "custom",
          "char_filter": [],
          "filter": ["lowercase"]
        }
      },
      "tokenizer": {
        "recid_tokenizer": {
          "type": "pattern",
          "pattern": "(\\d+)",
          "group": 1
        }
      }
    }
  }
}
```

**Comment by michamos** [comment]
I'm really surprised by the rather low accuracy and by some of the things you need to do to get acceptable results. Let's discuss it on Monday.

**Comment by michamos** [comment]
after a good night's sleep, I tried a different approach. Instead of trying to match raw affiliations with institutions, which is very complicated due to bad metadata, I tried matching raw affiliations with raw affiliations in literature records, and taking the corresponding ICN from the paper. It looks promising, I get 83% accuracy, although that number is a bit meaningless due to the affiliations I'm testing the approach with also being present in the database. For an independent test I'd need to run it on new affiliations coming from arXiv. The main issue is ambiguity when there are multiple raw affiliations for one author (as you don't know which correspond to which ICN), but it should be possible to exclude those. I've extended @MJedr's script and put it together with the results in the gist https://gist.github.com/michamos/eaa20e813a615f9c60ab79540f79c2de
